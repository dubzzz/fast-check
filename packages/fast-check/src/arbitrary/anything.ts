import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { anyArbitraryBuilder } from './_internals/builders/AnyArbitraryBuilder';
import type { ObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';
import { toQualifiedObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';

export type { ObjectConstraints };

/**
 * For any type of values
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(): Arbitrary<unknown>;
/**
 * For any type of values following the constraints defined by `settings`
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @example
 * ```typescript
 * // Using custom settings
 * fc.anything({
 *     key: fc.char(),
 *     values: [fc.integer(10,20), fc.constant(42)],
 *     maxDepth: 2
 * });
 * // Can build entries such as:
 * // - 19
 * // - [{"2":12,"k":15,"A":42}]
 * // - {"4":[19,13,14,14,42,11,20,11],"6":42,"7":16,"L":10,"'":[20,11],"e":[42,20,42,14,13,17]}
 * // - [42,42,42]...
 * ```
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(constraints: ObjectConstraints): Arbitrary<unknown>;
function anything(constraints?: ObjectConstraints): Arbitrary<unknown> {
  return anyArbitraryBuilder(toQualifiedObjectConstraints(constraints));
}
export { anything };
