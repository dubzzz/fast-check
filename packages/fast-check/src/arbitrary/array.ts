import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { ArrayArbitrary } from './_internals/ArrayArbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import {
  MaxLengthUpperBound,
  maxGeneratedLengthFromSizeForArbitrary,
} from './_internals/helpers/MaxLengthFromMinLength';
import type { DepthIdentifier } from './_internals/helpers/DepthContext';

/**
 * Constraints to be applied on {@link array}
 * @remarks Since 2.4.0
 * @public
 */
export interface ArrayConstraints {
  /**
   * Lower bound of the generated array size
   * @defaultValue 0
   * @remarks Since 2.4.0
   */
  minLength?: number;
  /**
   * Upper bound of the generated array size
   * @defaultValue 0x7fffffff â€” _defaulting seen as "max non specified" when `defaultSizeToMaxWhenMaxSpecified=true`_
   * @remarks Since 2.4.0
   */
  maxLength?: number;
  /**
   * Define how large the generated values should be (at max)
   *
   * When used in conjonction with `maxLength`, `size` will be used to define
   * the upper bound of the generated array size while `maxLength` will be used
   * to define and document the general maximal length allowed for this case.
   *
   * @remarks Since 2.22.0
   */
  size?: SizeForArbitrary;
  /**
   * When receiving a depth identifier, the arbitrary will impact the depth
   * attached to it to avoid going too deep if it already generated lots of items.
   *
   * In other words, if the number of generated values within the collection is large
   * then the generated items will tend to be less deep to avoid creating structures a lot
   * larger than expected.
   *
   * For the moment, the depth is not taken into account to compute the number of items to
   * define for a precise generate call of the array. Just applied onto eligible items.
   *
   * @remarks Since 2.25.0
   */
  depthIdentifier?: DepthIdentifier | string;
}

/**
 * Extra but internal constraints that can be passed to array.
 * This extra set is made of constraints mostly targets experimental and internal features not yet mature to be exposed.
 * @internal
 */
export interface ArrayConstraintsInternal<T> extends ArrayConstraints {
  /**
   * Extra user-definable and hardcoded values.
   * Each entry in the array could be used to build the final generated value outputed by the arbitrary of array on generate.
   * Each entry must have at least one element of type T into it.
   * Each T must be a value acceptable for the arbitrary passed to the array.
   */
  experimentalCustomSlices?: T[][];
}

/**
 * For arrays of values coming from `arb`
 *
 * @param arb - Arbitrary used to generate the values inside the array
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function array<T>(arb: Arbitrary<T>, constraints: ArrayConstraints = {}): Arbitrary<T[]> {
  const size = constraints.size;
  const minLength = constraints.minLength || 0;
  const maxLengthOrUnset = constraints.maxLength;
  const depthIdentifier = constraints.depthIdentifier;
  const maxLength = maxLengthOrUnset !== undefined ? maxLengthOrUnset : MaxLengthUpperBound;
  const specifiedMaxLength = maxLengthOrUnset !== undefined;
  const maxGeneratedLength = maxGeneratedLengthFromSizeForArbitrary(size, minLength, maxLength, specifiedMaxLength);
  const customSlices = (constraints as ArrayConstraintsInternal<T>).experimentalCustomSlices || [];
  return new ArrayArbitrary<T>(arb, minLength, maxGeneratedLength, maxLength, depthIdentifier, undefined, customSlices);
}
export { array };
