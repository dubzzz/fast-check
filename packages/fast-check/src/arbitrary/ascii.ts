import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';
import { indexToPrintableIndexMapper, indexToPrintableIndexUnmapper } from './_internals/mappers/IndexToPrintableIndex';

/**
 * For single ascii characters - char code between 0x00 (included) and 0x7f (included)
 * @remarks Since 0.0.1
 * @public
 */
export function ascii(): Arbitrary<string> {
  return buildCharacterArbitrary(0x00, 0x7f, indexToPrintableIndexMapper, indexToPrintableIndexUnmapper);
}
