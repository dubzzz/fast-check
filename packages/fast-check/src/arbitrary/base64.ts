import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';

/** @internal */
function base64Mapper(v: number) {
  if (v < 26) return v + 65; // A-Z
  if (v < 52) return v + 97 - 26; // a-z
  if (v < 62) return v + 48 - 52; // 0-9
  return v === 62 ? 43 : 47; // +/
}

/** @internal */
function base64Unmapper(v: number) {
  if (v >= 65 && v <= 90) return v - 65; // A-Z
  if (v >= 97 && v <= 122) return v - 97 + 26; // a-z
  if (v >= 48 && v <= 57) return v - 48 + 52; // 0-9
  return v === 43 ? 62 : v === 47 ? 63 : -1; // +/
}

/**
 * For single base64 characters - A-Z, a-z, 0-9, + or /
 * @remarks Since 0.0.1
 * @public
 */
export function base64(): Arbitrary<string> {
  return buildCharacterArbitrary(0, 63, base64Mapper, base64Unmapper);
}
