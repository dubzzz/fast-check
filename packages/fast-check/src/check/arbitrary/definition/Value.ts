import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
