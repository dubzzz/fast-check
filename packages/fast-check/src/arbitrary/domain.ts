import { array } from './array';
import {
  buildLowerAlphaArbitrary,
  buildLowerAlphaNumericArbitrary,
} from './_internals/builders/CharacterRangeArbitraryBuilder';
import { option } from './option';
import { stringOf } from './stringOf';
import { tuple } from './tuple';
import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { filterInvalidSubdomainLabel } from './_internals/helpers/InvalidSubdomainLabelFiIter';
import type { Size, SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import { resolveSize, relativeSizeToSize } from './_internals/helpers/MaxLengthFromMinLength';
import type { AdapterOutput } from './_internals/AdapterArbitrary';
import { adapter } from './_internals/AdapterArbitrary';
import { safeJoin, safeSlice, safeSplit, safeSubstring } from '../utils/globals';

/** @internal */
function toSubdomainLabelMapper([f, d]: [string, [string, string] | null]): string {
  return d === null ? f : `${f}${d[0]}${d[1]}`;
}
/** @internal */
function toSubdomainLabelUnmapper(value: unknown): [string, [string, string] | null] {
  if (typeof value !== 'string' || value.length === 0) {
    throw new Error('Unsupported');
  }
  if (value.length === 1) {
    return [value[0], null];
  }
  return [value[0], [safeSubstring(value, 1, value.length - 1), value[value.length - 1]]];
}

/** @internal */
function subdomainLabel(size: Size) {
  const alphaNumericArb = buildLowerAlphaNumericArbitrary([]);
  const alphaNumericHyphenArb = buildLowerAlphaNumericArbitrary(['-']);
  // Rq: maxLength = 61 because max length of a label is 63 according to RFC 1034
  //     and we add 2 characters to this generated value
  // According to RFC 1034 (confirmed by RFC 1035):
  //   <label>       ::= <letter> [ [ <ldh-str> ] <let-dig> ]
  //   <ldh-str>     ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
  //   <let-dig-hyp> ::= <let-dig> | "-"
  //   <letter>      ::= any one of the 52 alphabetic characters A through Z in upper case and a through z in lower case
  //   <digit>       ::= any one of the ten digits 0 through 9
  //   "The labels must follow the rules for ARPANET host names.  They must start with a letter, end with a letter or digit, and have as interior
  //    characters only letters, digits, and hyphen.  There are also some restrictions on the length.  Labels must be 63 characters or less."
  // But RFC 1123 relaxed the constraint:
  //   "The syntax of a legal Internet host name was specified in RFC-952[DNS:4]. One aspect of host name syntax is hereby changed: the
  //    restriction on the first character is relaxed to allow either a letter or a digit. Host software MUST support this more liberal syntax."
  return tuple(
    alphaNumericArb,
    option(tuple(stringOf(alphaNumericHyphenArb, { size, maxLength: 61 }), alphaNumericArb)),
  )
    .map(toSubdomainLabelMapper, toSubdomainLabelUnmapper)
    .filter(filterInvalidSubdomainLabel);
}

/** @internal */
function labelsMapper(elements: [string[], string]): string {
  return `${safeJoin(elements[0], '.')}.${elements[1]}`;
}
/** @internal */
function labelsUnmapper(value: unknown): [string[], string] {
  if (typeof value !== 'string') {
    throw new Error('Unsupported type');
  }
  const lastDotIndex = value.lastIndexOf('.');
  return [safeSplit(safeSubstring(value, 0, lastDotIndex), '.'), safeSubstring(value, lastDotIndex + 1)];
}

/** @internal */
function labelsAdapter(labels: [string[], string]): AdapterOutput<[string[], string]> {
  // labels[0].length is always >=1
  const [subDomains, suffix] = labels;
  let lengthNotIncludingIndex = suffix.length;
  for (let index = 0; index !== subDomains.length; ++index) {
    lengthNotIncludingIndex += 1 + subDomains[index].length;
    // Required by RFC 1034:
    //    To simplify implementations, the total number of octets that represent
    //    a domain name (i.e., the sum of all label octets and label lengths) is limited to 255.
    // It seems that this restriction has been relaxed in modern web browsers.
    if (lengthNotIncludingIndex > 255) {
      return { adapted: true, value: [safeSlice(subDomains, 0, index), suffix] };
    }
  }
  return { adapted: false, value: labels };
}

/**
 * Constraints to be applied on {@link domain}
 * @remarks Since 2.22.0
 * @public
 */
export interface DomainConstraints {
  /**
   * Define how large the generated values should be (at max)
   * @remarks Since 2.22.0
   */
  size?: Exclude<SizeForArbitrary, 'max'>;
}

/**
 * For domains
 * having an extension with at least two lowercase characters
 *
 * According to {@link https://www.ietf.org/rfc/rfc1034.txt | RFC 1034},
 * {@link https://www.ietf.org/rfc/rfc1035.txt | RFC 1035},
 * {@link https://www.ietf.org/rfc/rfc1123.txt | RFC 1123} and
 * {@link https://url.spec.whatwg.org/ | WHATWG URL Standard}
 *
 * @param constraints - Constraints to apply when building instances (since 2.22.0)
 *
 * @remarks Since 1.14.0
 * @public
 */
export function domain(constraints: DomainConstraints = {}): Arbitrary<string> {
  const resolvedSize = resolveSize(constraints.size);
  const resolvedSizeMinusOne = relativeSizeToSize('-1', resolvedSize);
  // A list of public suffixes can be found here: https://publicsuffix.org/list/public_suffix_list.dat
  // our current implementation does not follow this list and generate a fully randomized suffix
  // which is probably not in this list (probability would be low)
  const alphaNumericArb = buildLowerAlphaArbitrary([]);
  const publicSuffixArb = stringOf(alphaNumericArb, { minLength: 2, maxLength: 63, size: resolvedSizeMinusOne });
  return (
    // labels have between 1 and 63 characters
    // domains are made of dot-separated labels and have up to 255 characters so that are made of up-to 128 labels
    adapter(
      tuple(
        array(subdomainLabel(resolvedSize), { size: resolvedSizeMinusOne, minLength: 1, maxLength: 127 }),
        publicSuffixArb,
      ),
      labelsAdapter,
    ).map(labelsMapper, labelsUnmapper)
  );
}
