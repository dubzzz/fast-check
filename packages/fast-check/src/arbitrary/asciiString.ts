import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { array } from './array';
import { ascii } from './ascii';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings of {@link ascii}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function asciiString(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return array(ascii(), constraints).map(codePointsToStringMapper, codePointsToStringUnmapper);
}
