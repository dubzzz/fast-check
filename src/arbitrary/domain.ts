import { array } from './array';
import {
  buildLowerAlphaArbitrary,
  buildLowerAlphaNumericArbitrary,
} from './_internals/builders/CharacterRangeArbitraryBuilder';
import { option } from './option';
import { stringOf } from './stringOf';
import { tuple } from './tuple';
import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { filterInvalidSubdomainLabel } from './_internals/helpers/InvalidSubdomainLabelFiIter';

/** @internal */
function toSubdomainLabelMapper([f, d]: [string, [string, string] | null]): string {
  return d === null ? f : `${f}${d[0]}${d[1]}`;
}
/** @internal */
function toSubdomainLabelUnmapper(value: unknown): [string, [string, string] | null] {
  if (typeof value !== 'string' || value.length === 0) {
    throw new Error('Unsupported');
  }
  if (value.length === 1) {
    return [value[0], null];
  }
  return [value[0], [value.substring(1, value.length - 1), value[value.length - 1]]];
}

/** @internal */
function subdomainLabel() {
  const alphaNumericArb = buildLowerAlphaNumericArbitrary([]);
  const alphaNumericHyphenArb = buildLowerAlphaNumericArbitrary(['-']);
  // Rq: maxLength = 61 because max length of a label is 63 according to RFC 1034
  //     and we add 2 characters to this generated value
  return tuple(alphaNumericArb, option(tuple(stringOf(alphaNumericHyphenArb, { maxLength: 61 }), alphaNumericArb)))
    .map(toSubdomainLabelMapper, toSubdomainLabelUnmapper)
    .filter(filterInvalidSubdomainLabel);
}

/** @internal */
function labelsMapper(elements: [string[], string]): string {
  return `${elements[0].join('.')}.${elements[1]}`;
}
/** @internal */
function labelsUnmapper(value: unknown): [string[], string] {
  if (typeof value !== 'string') {
    throw new Error('Unsupported type');
  }
  const lastDotIndex = value.lastIndexOf('.');
  return [value.substring(0, lastDotIndex).split('.'), value.substring(lastDotIndex + 1)];
}

/**
 * For domains
 * having an extension with at least two lowercase characters
 *
 * According to {@link https://www.ietf.org/rfc/rfc1034.txt | RFC 1034},
 * {@link https://www.ietf.org/rfc/rfc1035.txt | RFC 1035},
 * {@link https://www.ietf.org/rfc/rfc1123.txt | RFC 1123} and
 * {@link https://url.spec.whatwg.org/ | WHATWG URL Standard}
 *
 * @remarks Since 1.14.0
 * @public
 */
export function domain(): Arbitrary<string> {
  const alphaNumericArb = buildLowerAlphaArbitrary([]);
  // A list of public suffixes can be found here: https://publicsuffix.org/list/public_suffix_list.dat
  // our current implementation does not follow this list and generate a fully randomized suffix
  // which is probably not in this list (probability would be low)
  const publicSuffixArb = stringOf(alphaNumericArb, { minLength: 2, maxLength: 10 });
  return (
    tuple(array(subdomainLabel(), { minLength: 1, maxLength: 5 }), publicSuffixArb)
      .map(labelsMapper, labelsUnmapper)
      // Required by RFC 1034:
      //    To simplify implementations, the total number of octets that represent
      //    a domain name (i.e., the sum of all label octets and label lengths) is limited to 255.
      // It seems that this restriction has been relaxed in modern web browsers.
      .filter((d) => d.length <= 255)
  );
}
