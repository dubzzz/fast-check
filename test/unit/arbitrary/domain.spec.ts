import fc from '../../../lib/fast-check';
import { domain } from '../../../src/arbitrary/domain';
import { Value } from '../../../src/check/arbitrary/definition/Value';

import {
  assertProduceSameValueGivenSameSeed,
  assertProduceCorrectValues,
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
} from './__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('domain (integration)', () => {
  const isValidDomain = (t: string) => {
    // According to https://www.ietf.org/rfc/rfc1034.txt
    // <domain> ::= <subdomain> | " "
    // <subdomain> ::= <label> | <subdomain> "." <label>
    // <label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]
    // <ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
    // <let-dig-hyp> ::= <let-dig> | "-"
    // <let-dig> ::= <letter> | <digit>
    // Relaxed by https://www.ietf.org/rfc/rfc1123.txt
    // allowing first character of subdomain to be a digit
    const rfc1123SubDomain = /^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$/;
    return t.split('.').every((sd) => rfc1123SubDomain.test(sd) && sd.length <= 63) && t.length <= 255;
  };

  const isValidDomainWithExtension = (t: string) => {
    const subdomains = t.split('.');
    return isValidDomain(t) && subdomains.length >= 2 && /^[a-z]{2,}$/.test(subdomains[subdomains.length - 1]);
  };

  const isCorrect = isValidDomainWithExtension;

  const domainBuilder = () => domain();

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(domainBuilder);
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(domainBuilder, isCorrect);
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(domainBuilder);
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(domainBuilder);
  });

  it.each`
    rawValue
    ${'domain.com'}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = domain();
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
