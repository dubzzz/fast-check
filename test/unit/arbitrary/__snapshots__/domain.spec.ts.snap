// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`domain (integration) should be able to shrink domain.com 1`] = `
"\\"domain.com\\"
├> \\"aomain.com\\"
|  ├> \\"a.com\\"
|  |  ├> \\"a.om\\"
|  |  |  ├> \\"a.am\\"
|  |  |  ├> \\"a.hm\\"
|  |  |  |  ├> \\"a.em\\"
|  |  |  |  |  ├> \\"a.cm\\"
|  |  |  |  |  |  └> \\"a.bm\\"
|  |  |  |  |  |     └> \\"a.am\\"
|  |  |  |  |  └> \\"a.dm\\"
|  |  |  |  |     └> \\"a.cm\\"
|  |  |  |  |        ├> \\"a.am\\"
|  |  |  |  |        └> \\"a.bm\\"
|  |  |  |  |           └> \\"a.am\\"
|  |  |  |  └> \\"a.gm\\"
|  |  |  |     └> \\"a.fm\\"
|  |  |  |        └> \\"a.em\\"
|  |  |  |           ├> \\"a.am\\"
|  |  |  |           ├> \\"a.cm\\"
|  |  |  |           |  └> \\"a.bm\\"
|  |  |  |           |     └> \\"a.am\\"
|  |  |  |           └> \\"a.dm\\"
|  |  |  |              └> \\"a.cm\\"
|  |  |  |                 ├> \\"a.am\\"
|  |  |  |                 └> \\"a.bm\\"
|  |  |  |                    └> \\"a.am\\"
|  |  |  ├> \\"a.lm\\"
|  |  |  |  ├> \\"a.jm\\"
|  |  |  |  |  └> \\"a.im\\"
|  |  |  |  |     └> \\"a.hm\\"
|  |  |  |  |        ├> \\"a.am\\"
|  |  |  |  |        ├> \\"a.em\\"
|  |  |  |  |        |  ├> \\"a.cm\\"
|  |  |  |  |        |  |  └> \\"a.bm\\"
|  |  |  |  |        |  |     └> \\"a.am\\"
|  |  |  |  |        |  └> \\"a.dm\\"
|  |  |  |  |        |     └> \\"a.cm\\"
|  |  |  |  |        |        ├> \\"a.am\\"
|  |  |  |  |        |        └> \\"a.bm\\"
|  |  |  |  |        |           └> \\"a.am\\"
|  |  |  |  |        └> \\"a.gm\\"
|  |  |  |  |           └> \\"a.fm\\"
|  |  |  |  |              └> \\"a.em\\"
|  |  |  |  |                 ├> \\"a.am\\"
|  |  |  |  |                 ├> \\"a.cm\\"
|  |  |  |  |                 |  └> \\"a.bm\\"
|  |  |  |  |                 |     └> \\"a.am\\"
|  |  |  |  |                 └> \\"a.dm\\"
|  |  |  |  |                    └> \\"a.cm\\"
|  |  |  |  |                       ├> \\"a.am\\"
|  |  |  |  |                       └> \\"a.bm\\"
|  |  |  |  |                          └> \\"a.am\\"
|  |  |  |  └> \\"a.km\\"
|  |  |  |     └> \\"a.jm\\"
|  |  |  |        ├> \\"a.am\\"
|  |  |  |        ├> \\"a.fm\\"
|  |  |  |        |  ├> \\"a.dm\\"
|  |  |  |        |  |  └> \\"a.cm\\"
|  |  |  |        |  |     └> \\"a.bm\\"
|  |  |  |        |  |        └> \\"a.am\\"
|  |  |  |        |  └> \\"a.em\\"
|  |  |  |        |     └> \\"a.dm\\"
|  |  |  |        |        ├> \\"a.am\\"
|  |  |  |        |        └> \\"a.cm\\"
|  |  |  |        |           └> \\"a.bm\\"
|  |  |  |        |              └> \\"a.am\\"
|  |  |  |        ├> \\"a.hm\\"
|  |  |  |        |  └> \\"a.gm\\"
|  |  |  |        |     └> \\"a.fm\\"
|  |  |  |        |        ├> \\"a.am\\"
|  |  |  |        |        ├> \\"a.dm\\"
|  |  |  |        |        |  └> \\"a.cm\\"
|  |  |  |        |        |     └> \\"a.bm\\"
|  |  |  |        |        |        └> \\"a.am\\"
|  |  |  |        |        └> \\"a.em\\"
|  |  |  |        |           └> \\"a.dm\\"
|  |  |  |        |              ├> \\"a.am\\"
|  |  |  |        |              └> \\"a.cm\\"
|  |  |  |        |                 └> \\"a.bm\\"
|  |  |  |        |                    └> \\"a.am\\"
|  |  |  |        └> \\"a.im\\"
|  |  |  |           └> \\"a.hm\\"
|  |  |  |              ├> \\"a.am\\"
|  |  |  |              ├> \\"a.em\\"
|  |  |  |              |  ├> \\"a.cm\\"
|  |  |  |              |  |  └> \\"a.bm\\"
|  |  |  |              |  |     └> \\"a.am\\"
|  |  |  |              |  └> \\"a.dm\\"
|  |  |  |              |     └> \\"a.cm\\"
|  |  |  |              |        ├> \\"a.am\\"
|  |  |  |              |        └> \\"a.bm\\"
|  |  |  |              |           └> \\"a.am\\"
|  |  |  |              └> \\"a.gm\\"
|  |  |  |                 └> \\"a.fm\\"
|  |  |  |                    └> \\"a.em\\"
|  |  |  |                       ├> \\"a.am\\"
|  |  |  |                       ├> \\"a.cm\\"
|  |  |  |                       |  └> \\"a.bm\\"
|  |  |  |                       |     └> \\"a.am\\"
|  |  |  |                       └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
